File: epitaxy.c
//    Organic Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
//    model for organic solar cells. 
//    Copyright (C) 2012 Roderick C. I. MacKenzie
//
//	roderick.mackenzie@nottingham.ac.uk
//	www.roderickmackenzie.eu
//	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License along
//    with this program; if not, write to the Free Software Foundation, Inc.,
//    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "epitaxy.h"
#include "inp.h"
#include "util.h"
#include "true_false.h"

void epitaxy_load(struct epitaxy *in, char *file)
{
	int i;
	char dos_file[20];
	char pl_file[20];
	struct inp_file inp;
	in->electrical_layers = 0;

	inp_init(&inp);
	inp_load(&inp, file);
	inp_check(&inp, 1.1);
	inp_reset_read(&inp);
	inp_get_string(&inp);
	sscanf(inp_get_string(&inp), "%d", &(in->layers));

	if (in->layers > 20) {
		ewe("Too many material layers\n");
	}

	if (in->layers < 1) {
		ewe("No material layers\n");
	}

	for (i = 0; i < in->layers; i++) {
		inp_get_string(&inp);
		sscanf(inp_get_string(&inp), "%le", &(in->width[i]));
		in->width[i] = fabs(in->width[i]);
		strcpy(in->mat_file[i], inp_get_string(&inp));
		strcpy(dos_file, inp_get_string(&inp));
		strcpy(pl_file, inp_get_string(&inp));

		char temp[20];
		if (strcmp(dos_file, "none") != 0) {
			strcpy(temp, dos_file);
			strcat(temp, ".inp");
			in->electrical_layer[i] = TRUE;
			if (inp_isfile(temp) != 0) {
				ewe("dos file %s does not exist", temp);
			}
			strcpy(in->dos_file[in->electrical_layers], dos_file);

			strcpy(temp, pl_file);
			strcat(temp, ".inp");
			if (inp_isfile(temp) != 0) {
				ewe("pl file %s does not exist", temp);
			}
			strcpy(in->pl_file[in->electrical_layers], pl_file);

			in->electrical_layers++;
		} else {
			in->electrical_layer[i] = FALSE;
		}

	}

	char *ver = inp_get_string(&inp);
	if (strcmp(ver, "#ver") != 0) {
		ewe("No #ver tag found in file\n");
	}

	inp_free(&inp);
}

double epitaxy_get_electrical_length(struct epitaxy *in)
{
	int i = 0;
	double tot = 0.0;

	for (i = 0; i < in->layers; i++) {
		printf("rod>>>%d\n", in->electrical_layer[i]);
		if (in->electrical_layer[i] == TRUE) {
			tot += in->width[i];
		}
	}

	return tot;
}

double epitaxy_get_optical_length(struct epitaxy *in)
{
	int i = 0;
	double tot = 0.0;

	for (i = 0; i < in->layers; i++) {
		tot += in->width[i];
	}

	return tot;
}

int epitaxy_get_optical_material_layer(struct epitaxy *in, double pos)
{
	int i = 0;
	double layer_end = 0.0;
	for (i = 0; i < in->layers; i++) {
		layer_end += in->width[i];

		if (pos < layer_end) {
			return i;
		}

	}

	return -1;
}

int epitaxy_get_electrical_material_layer(struct epitaxy *in, double pos)
{
	int i = 0;
	double layer_end = 0.0;
	int electrical_layer = 0;

	for (i = 0; i < in->layers; i++) {
		if (in->electrical_layer[i] == TRUE) {
			layer_end += in->width[i];

			if (pos < layer_end) {
				return electrical_layer;
			}
			electrical_layer++;
		}

	}

	return -1;
}

double epitaxy_get_device_start(struct epitaxy *in)
{
	int i = 0;
	double pos = 0.0;
	for (i = 0; i < in->layers; i++) {

		if (in->electrical_layer[i] == TRUE) {
			return pos;
		}
		pos += in->width[i];

	}

	return -1;
}

double epitaxy_get_device_start_i(struct epitaxy *in)
{
	int i = 0;
	for (i = 0; i < in->layers; i++) {

		if (in->electrical_layer[i] == TRUE) {
			return i;
		}

	}

	return -1;
}

File: gui/epitaxy.py
#    Organic Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
#    model for organic solar cells. 
#    Copyright (C) 2012 Roderick C. I. MacKenzie
#
#	roderick.mackenzie@nottingham.ac.uk
#	www.opvdm.com
#	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License v2.0, as published by
#    the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this program; if not, write to the Free Software Foundation, Inc.,
#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


import pygtk
pygtk.require('2.0')
import gtk
import sys
import os
import shutil
from inp import inp_update_token_value
from inp import inp_write_lines_to_file
from inp import inp_load_file
from inp import inp_search_token_value
from scan_item import scan_item
from scan_item import scan_item_add
from cal_path import find_data_file
import glob
from util import time_with_units
from plot_widget import plot_widget
from util_zip import zip_get_data_file
from window_list import windows
from plot_state import plot_state
from plot_io import plot_load_info
from cal_path import get_exe_command


layers=0
electrical_layers=0
width=[]
mat_file=[]
electrical_layer=[]
pl_file=[]
name=[]

def epitaxy_load():
	lines=[]
	global layers
	global electrical_layers
	global width
	global mat_file
	global electrical_layer
	global pl_file
	global name

	layers=0
	electrical_layers=0
	width=[]
	mat_file=[]
	electrical_layer=[]
	pl_file=[]
	name=[]

	if inp_load_file(lines,"epitaxy.inp")==True:
		pos=0
		pos=pos+1

		for i in range(0, int(lines[pos])):
			pos=pos+1
			#label=lines[pos]				#read label

			pos=pos+1
			name.append(lines[pos])

			pos=pos+1
			width.append(float(lines[pos]))

			pos=pos+1
			mat_file.append(lines[pos])

			pos=pos+1
			electrical_layer.append(lines[pos])		#value

			if lines[pos]!="none":
				electrical_layers=electrical_layers+1

			pos=pos+1
			pl_file.append(lines[pos])		#value

			layers=layers+1

def epitay_get_next_dos():
	global electrical_layer
	for i in range(0,20):
		name="dos"+str(i)
		if electrical_layer.count(name)==0:
			return name

def epitay_get_next_pl():
	global pl_file
	for i in range(0,20):
		name="pl"+str(i)
		if pl_file.count(name)==0:
			return name

def epitaxy_load_from_arrays(in_name,in_width,in_material,in_dos_layer,in_pl_file):
	lines=[]
	global layers
	global electrical_layers
	global width
	global mat_file
	global electrical_layer
	global pl_file
	global name

	layers=0
	electrical_layers=0
	width=[]
	mat_file=[]
	electrical_layer=[]
	pl_file=[]
	name=[]

	for i in range(0, len(in_width)):

		name.append(in_name[i])

		width.append(float(in_width[i]))

		mat_file.append(in_material[i])

		electrical_layer.append(in_dos_layer[i])		#value

		pl_file.append(in_pl_file[i])			#value

		if in_dos_layer[i]!="none":
			electrical_layers=electrical_layers+1


		layers=layers+1

def epitaxy_save():
	global layers
	global electrical_layers
	global width
	global mat_file
	global electrical_layer
	global pl_file
	global name

	dos_text=""
	lines=[]
	lines.append("#layers")
	lines.append(str(layers))

	layer=0
	for i in range(0,layers):
		lines.append("#layer"+str(layer))
		lines.append(str(name[i]))
		lines.append(str(width[i]))
		lines.append(mat_file[i])
		lines.append(electrical_layer[i])
		lines.append(pl_file[i])
		layer=layer+1

	lines.append("#ver")			
	lines.append("1.2")			
	lines.append("#end")

	#print lines
	inp_write_lines_to_file(os.path.join(os.getcwd(),"epitaxy.inp"),lines)

def epitaxy_get_dos_files():
	global electrical_layer
	dos_file=[]
	for i in range(0,len(electrical_layer)):
		if electrical_layer[i]!="none":
			dos_file.append(electrical_layer[i])

	return dos_file

def epitaxy_get_layers():
	global layers
	return layers

def epitaxy_get_width(i):
	global width
	return width[i]

def epitaxy_get_electrical_layer(i):
	global electrical_layer
	#print electrical_layer,i
	return electrical_layer[i]

def epitaxy_get_mat_file(i):
	global mat_file
	return mat_file[i]

def epitaxy_get_pl_file(i):
	global pl_file
	return pl_file[i]

def epitaxy_get_name(i):
	global name
	return name[i]

